<?xml version="1.0" encoding="UTF-8"?>
<launch>
  <arg name="model" default="$(find sim2_lecture)/xacro/odm_try_wheel1.xacro" />
  <arg name="rvizconfig" default="$(find sim2_lecture)/rviz/odm_display.rviz" />  
  <param name="robot_description" command="$(find xacro)/xacro $(arg model) --inorder"/>

  <include file="$(find gazebo_ros)/launch/empty_world.launch">
    <arg name="paused" value="false"/>
    <arg name="use_sim_time" value="true"/>
    <arg name="gui" value="true"/>
    <arg name="headless" value="false"/>
    <arg name="debug" value="false"/>
  </include>

  <node name="spawn_urdf" pkg="gazebo_ros" type="spawn_model" args="-param robot_description -urdf -model odm_robot" />

  <rosparam command="load" file="$(find sim2_lecture)/config/odm_move_controller.yaml" ns="odm_robot"/>  
  <rosparam command="load" file="$(find sim2_lecture)/config/odm_joint_state_controller.yaml" ns="odm_robot"/>  
  <node name="controller_spawner" pkg="controller_manager"
        type="spawner" respawn="false"
        output="screen" ns="/odm_robot"
        args="joint_controller0 joint_controller1 joint_controller2 joint_state_controller"/>

  <node name="sim2_omni_commander" pkg="sim2_lecture" type="sim2_omni_commander" output="screen">
    <param name="wheel_base" value="0.0972"/>
    <param name="wheel_radius" value="0.019"/>
    <param name="wheel0" value="${radians(60)}"/>
    <param name="wheel1" value="${radians(180)}"/>
    <param name="wheel2" value="${radians(300)}"/>
    <remap from="wheel0" to="/odm_robot/joint_controller0/command" />
    <remap from="wheel1" to="/odm_robot/joint_controller1/command" />
    <remap from="wheel2" to="/odm_robot/joint_controller2/command" />
  </node>

  <node name="sim2_omni_odometry" pkg="sim2_lecture" type="sim2_omni_odometry" output="screen">
    <param name="wheel_base" value="0.0972"/>
    <param name="wheel_radius" value="0.019"/>
    <param name="frame_id" value="odm_robot/odom"/>
    <remap from="wheel0/state" to="/odm_robot/joint_controller0/state" />
    <remap from="wheel1/state" to="/odm_robot/joint_controller1/state" />
    <remap from="wheel2/state" to="/odm_robot/joint_controller2/state" />
  </node>

  <node name="sim1_model_tf_publisher" pkg="sim1_lecture" type="sim1_model_tf_publisher">
    <param name="model_name" value="odm_robot" />
    <param name="base_frame" value="base_link" />
    <param name="world_frame" value="odom" />
  </node>

  <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher" output="screen" ns="/odm_robot"/>
  <node name="rviz" pkg="rviz" type="rviz" args="-d $(arg rvizconfig)" required="true" />
</launch>
